### ISSUE TITLE
Changing a resource does not clear the cache for all the transforms

### ISSUE THREAD

**Issue by ajoliveau** [workingconcept/cloudflare-craft-plugin#67]
Running on Craft 4.5.11.1 with the Cloudflare pluggin v2.0.2

I set the plugin up to clear the Cloudflare cache when a resource is updated. The main resource image is correctly purged, but all the transforms (crops, conversion, etc) is not.
I usually create transforms to have images available as webp, to resize them so they're used in srcsets, and to crop them depending on the focal point.

I shoud specify that they are not named assets, they're dynamically generated at runtime in Twig.
So if my "source" asset's url is `uploads/spectacles/show_-NWI_fEElSZn3sx6wBmf_-full.jpg`, I have transforms named like `uploads/spectacles/_600x450_crop_center-center_85_none/2378715/show_-NWI_fEElSZn3sx6wBmf_-full.webp`

If I go to the Utilities and manually purge the transform url, it works correctly.

I can try to build my own purging logic using your API methods, but shouldn't this be part of the resource purging feature ?

![image](https://github.com/putyourlightson/craft-cloudflare/assets/3447717/a162a2c1-1539-4a14-b772-0acb0d95d1a0)

**Comment by bencroker** [comment]
I understand how you might think this should be automatically handled by the plugin, but it has no way of knowing about every existing image transform, especially since they can be create on-the-fly in the templates, so I’m not sure how to best handle this.

**Comment by ajoliveau** [comment]
Yep I understand it's a specific use-case, but is this one that is really so rare ?
I actually researched this, and it seems that Craft actually keeps track of all the transforms, even unnamed ones, in a specific DB table called `imagetransformindex`

![image](https://github.com/putyourlightson/craft-cloudflare/assets/3447717/0d3cb5f6-26e8-4485-b275-62df5bab74bf)


Craft has a built-in logic to, when an asset is changed or deleted, find all the transforms associated to this asset and delete them in the filesystem. You can see the method that's called here : https://docs.craftcms.com/api/v4/craft-imagetransforms-imagetransformer.html#method-getallcreatedtransformsforasset

Unfortunately all this logic is private, and there's no way to reuse their functions and methods, but what I managed to do is to listen on the `EVENT_BEFORE_INVALIDATE_ASSET_TRANSFORMS` event, and manually copy all the code to get the list of transforms when Craft would delete them. Then I just call `purgeUrls()` on them.

Here is the code inside my module : 

```
Event::on(
    ImageTransforms::class,
    ImageTransforms::EVENT_BEFORE_INVALIDATE_ASSET_TRANSFORMS,
    function($event) {
        $asset = $event->asset;
        $urlsToInvalidate = [];

        $transformIndexes = (new Query())
        ->select([
            'filename',
            'transformString',
        ])
        ->from([Table::IMAGETRANSFORMINDEX])
        ->where(['assetId' => $asset->id])
        ->all();

        $rootUrl = rtrim($asset->getVolume()->getTransformFs()->getRootUrl(), '/');
        $basePath = $asset->getVolume()->transformSubpath;
        $basePath = StringHelper::removeRight($basePath, '/');
        $basePath = ($basePath ? $basePath . DIRECTORY_SEPARATOR : '') . $asset->folderPath;

        foreach ($transformIndexes as $key => $transformIndex) {

            $subfolder = $transformIndex['transformString'];
            if (!empty($transformIndex['filename']) && $transformIndex['filename'] !== $asset->getFilename()) {
                $subfolder .= DIRECTORY_SEPARATOR . $asset->id;
            }

            $filename = $transformIndex['filename'] ?: $asset->getFilename();

            $path = $basePath . DIRECTORY_SEPARATOR . $subfolder . DIRECTORY_SEPARATOR . $filename;
            
            $url = $rootUrl . $path;

            $urlsToInvalidate[] = $url;
        }
        
        $return = Cloudflare::getInstance()->api->purgeUrls($urlsToInvalidate);
        Craft::info("Purged urls : ", 'custom-cf-purge');
        Craft::info($urlsToInvalidate, 'custom-cf-purge');
        Craft::info($return, 'custom-cf-purge');

    }    
);
```

It seems to work fine ! I still think this feature would be inside the scope of the plugin, it's nice to know it's doable at least :)

**Comment by bencroker** [comment]
You’re right, this does actually seem much more feasible given that Craft tracks all image transforms. Thanks for sharing your code, I’ll look into what it would take to add this to the plugin.

**Comment by ajoliveau** [comment]
Glad to hear that, thank you ! I can turn my code snippet above into a pull request if that can help.

**Comment by bencroker** [comment]
That’s ok, it’s a good starting point that I can reference but not yet sure how I’d integrate this.

**Comment by bencroker** [comment]
So after exploring, it turns out that we’re already doing something similar in Blitz: 
https://github.com/putyourlightson/craft-blitz/blob/3d838a8332327c093821ddd5247621ee59dab5a8/src/helpers/SiteUriHelper.php#L291-L312

The main difference is that Blitz is specifically tracking changes to assets based on modified images. So I’ve gone with your suggested event, which takes effect when Assets are selected as purgeable elements, in https://github.com/putyourlightson/craft-cloudflare/commit/9a683fa2b6774603124d15f18520c60f1bf50460

You can test this by running `composer require "putyourlightson/craft-cloudflare:dev-develop as 2.1.0"`.

**Comment by bencroker** [comment]
Have you had a chance to test this yet, @ajoliveau?

**Comment by bencroker** [comment]
Released in [2.1.0](https://github.com/putyourlightson/craft-cloudflare/releases/tag/2.1.0).
